### ISSUE TITLE
[BUG] Invalid groupPrincipals in token objects after 2.6.2 to 2.6.6 upgrade

### ISSUE THREAD

**Issue by samjustus** [rancher/rancher#39107]
SURE-5050

**Rancher Server Setup**
Rancher version: 2.6.6
Rancher Cluster:
Rancher version: 2.6.6
Number of nodes: 3
Node OS version: Linux

Downstream Cluster:
Number of Downstream clusters: 3
Node OS: Linux
RKE/RKE2/K3S version: RKE

**Describe the bug**
While creating any Token , there is no Validation happening by default and groupPrincipals is setting as null. When trying to edit that token , it's not allowing to save the token yaml.

========

1. tokens.management.cattle.io "token-8q2tp" was not valid:
2. * <nil>: Invalid value: "The edited file failed validation": unknown object type "nil" in Token.groupPrincipals
=========

**To Reproduce**
Rancher 2.6.6 , Go to Local cluster

1. Create a token ( token-gtxs8 ) either using the kubectl or from UI ( Create API Key from Local cluster)

2. Try to edit the description of token using ; $ kubectl edit token token-gtxs8

3. It won't allow to edit the token and throw the below error.

====

tokens.management.cattle.io "token-gtxs8" was not valid:
* <nil>: Invalid value: "The edited file failed validation": unknown object type "nil" in Token.groupPrincipals
====
4. If you change the groupPrincipals from null to [] as below , and try to edit the token its working

groupPrincipals: []

==
> kubectl edit token token-gtxs8
token.management.cattle.io/token-gtxs8 edited
==

The main concern is its not validating the CRD schema when creating a token first time but its validating while editing the token. Its making lot of manual effort

**Result**
groupPrincipals is set to null while creating a token and there is no initial validation happening, this violates against the CRD schema.

**Expected Result**
While creating a token and there should be the initial validation happening

**Screenshots**
see jira

**Additional context**
What is the workaround:

change the groupPrincipals from null to [] as below , and try to edit the token its working

==
apiVersion: management.cattle.io/v3
authProvider: local
current: false
description: lingo
expired: false
expiresAt: ""
groupPrincipals: []
isDerived: true
kind: Token

**Comment by crobby** [comment]
I've had a quick look at this today.  I think it's a simple fix to have groupPrincipals set correctly on create (currently, it's just absent in the code block that does the create).  I'm still digging into why it's not validated on create, but it is validated on update.

**Comment by crobby** [comment]
https://github.com/rancher/rancher/pull/39147

**Comment by samjustus** [comment]
@crobby moving this to 2.7.1 - I don't think we'll need a backport?

**Comment by brudnak** [comment]
# Reopened
> Reopened based on an upgrade test case. Please see the additional tests in the table at the bottom

## Reproduction Environment


| Component                        | Version / Type                          |
| -------------------------------- | --------------------------------------- |
| Rancher version                  | 2.6.9                                   |
| Installation option              | Helm (high availability)                |
| If Helm Chart k8s cluster        | v1.24.7+k3s1                            |
| Cert Details                     | Let's Encrypt / `traefik`               |
| k3s ha external db               | Aurora MySQL 5.7.mysql_aurora.2.10.2    |
| Helm version                     | v2.16.8-rancher2                        |
| Downstream cluster type          | Not applicable                          |
| Downstream K8s version           | Not applicable                          |
| Authentication providers enabled | local, GitHub                           |
| Logged in user role              | Administrator, Standard User            |
| Browser type                     | Google Chrome                           |
| Browser version                  | 108.0.5359.94 (Official Build) (x86_64) |

<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

>K3s install command for node one. Replace values in <>

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=SECRET --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```

>K3s install command for node two. Replace values in <>. Main difference here is the --token

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=<TOKEN_FROM_NODE_ONE> --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```


</details>


## Reproduction steps

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click at top right of screen `user icon` >> `Account & API Keys` >> `Create API Key`
3. Fill out the `API Key: Create` form
   1. Description >> `TEST_TOKEN`
   2. Scope >> `No Scope`
   3. Automatically expire >> `Never`
4. Using kubectl against the local cluster run these commands
5. `kubectl get tokens`
6. From the list notice the token you just created, now run
7. `kubectl edit token token-dh4bb`
8. Use the VIM editor to change the description: of the token from
   - `TEST_TOKEN` >> `UPDATED_TOKEN`
   - in the VIM editor, `esc`, `shift` + `:`, `wq`, `enter`
9. Updating the token fails with the following error in the VIM editor `unknown object type "nil" in Token.groupPrincipals`

```yaml
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# tokens.management.cattle.io "token-dh4bb" was not valid:
# * <nil>: Invalid value: "The edited file failed validation": unknown object type "nil" in Token.groupPrincipals
#
```

## Additional Info

## RESULTS

#### ✅ Expected
Expected to be able to edit the token with kubectl without error

#### ❌ Actual
Getting the following error when trying to updated the token description: `unknown object type "nil" in Token.groupPrincipals`

#### Workaround
For the yaml line `groupPrincipals: null` change it to >> `groupPrincipals: []`

<hr>

## Validation Environment


| Component                        | Version / Type                                                                     |
| -------------------------------- | ---------------------------------------------------------------------------------- |
| Rancher version                  | v2.7-44d3d40c6dd9acd0aae257ea0f80ea9eab743784-head                                 |
| Rancher commit link              | https://github.com/rancher/rancher/commit/44d3d40c6dd9acd0aae257ea0f80ea9eab743784 |
| Installation option              | Helm (high availability)                                                           |
| If Helm Chart k8s cluster        | v1.24.7+k3s1                                                                       |
| Cert Details                     | Let's Encrypt / `traefik`                                                          |
| k3s ha external db               | Aurora MySQL 5.7.mysql_aurora.2.10.2                                               |
| Helm version                     | v2.16.8-rancher2                                                                   |
| Downstream cluster type          | RKE1 Linode                                                                        |
| Downstream K8s version           | v1.24.8-rancher1-1                                                                 |
| Authentication providers enabled | local, GitHub                                                                      |
| Logged in user role              | Administrator, Standard User                                                       |
| Browser type                     | Google Chrome                                                                      |
| Browser version                  | 108.0.5359.94 (Official Build) (x86_64)                                            |


<details>
    <summary>🚨 Additional Reproduction Setup Details: Click to Expand</summary>

>K3s install command for node one. Replace values in <>

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=SECRET --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```

>K3s install command for node two. Replace values in <>. Main difference here is the --token

```shell
curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=<K3S_VERSION> sh -s - server --token=<TOKEN_FROM_NODE_ONE> --datastore-endpoint=<DB_ENDPOINT> --tls-san <TLD> --node-external-ip <IP>
```


</details>

## Validation steps

1. Starting from the **default** Rancher homepage `/dashboard/home`
2. Click at top right of screen `user icon` >> `Account & API Keys` >> `Create API Key`
3. Fill out the `API Key: Create` form
   1. Description >> `TEST_TOKEN`
   2. Scope >> `No Scope`
   3. Automatically expire >> `Never`
4. Using kubectl against the local cluster run these commands
5. `kubectl get tokens`
6. From the list notice the token you just created, now run
7. `kubectl edit token token-dh4bb`
8. Use the VIM editor to change the description: of the token from
   - `TEST_TOKEN` >> `UPDATED_TOKEN`
   - in the VIM editor, `esc`, `shift` + `:`, `wq`, `enter`
9. Updating the token description was successful

## Additional Info

## RESULTS

#### ✅ Expected
To be able to edit the token without needing to set the group principal to `[]`

#### ✅ Actual
Was able to successfully just update the token description without changing other fields


## 📋 QA Test Plan

<table>
    <tbody>
        <tr>
            <th>Scenario</th>
            <th>Steps</th>
            <th>Expected</th>
            <th>Actual & Pass/Fail</th>
        </tr>
        <tr>
            <td>Upgrade Rancher from v2.6.9 to v2.7-head then create and edit new tokens</td>
            <td>
                <ol>
                    <li>Start with Rancher 2.6.9</li>
                    <li>Upgrade Rancher to v2.7-head</li>
                    <li>Create a new API token</li>
                    <li>Edit the token with kubectl edit token $token_name</li>
                    <li>Change the description of the token</li>
                    <li>Update token is successful</li>
                </ol>
            </td>
            <td>Expected: Expected to be able to edit newly created tokens without the issue of group principal being nil</td>
            <td>Actual: Was successfully able to edit the newly created token after upgrade<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
        <tr>
            <td>As an admin, edit a token created by a standard user with GitHub auth</td>
            <td>
                <ol>
                    <li>Setup GitHub auth</li>
                    <li>Login as a GitHub standard user</li>
                    <li>Create an RKE1 Linode downstream cluster</li>
                    <li>Create an API token</li>
                    <li>Log out as the standard user</li>
                    <li>Login as an Administrator</li>
                    <li>Edit the token with kubectl edit token $token_name</li>
                    <li>Update the token description</li>
                    <li>Successfully able to edit the token description</li>
                </ol>
            </td>
            <td>Expected: As an administrator to still be able to edit a token from a standard user with an auth provider</td>
            <td>Actual: As an administrator was still able to edit the token from a standard user with an auth provider<br><br>Pass/Fail: ✅<code>PASS</code></td>
        </tr>
        <tr>
            <td>Starting on Rancher 2.6.9 create tokens as admin then upgrade to Rancher v2.7-head</td>
            <td>
                <ol>
                    <li>Starting with Rancher 2.6.9 create a token TOKEN_PRE_UPGRADE</li>
                    <li>Upgrade Rancher from 2.6.9 to v2.7-head</li>
                    <li>Edit the existing token that was created TOKEN_PRE_UPGRADE</li>
                    <li>With kubectl edit token $token_name</li>
                    <li>group principals is set to null on the old token that was brought over: groupPrincipals: null</li>
                    <li>When trying to edit solely the description getting the error</li>
                    <li>The edited file failed validation": unknown object type "nil" in Token.groupPrincipals</li>
                </ol>
            </td>
            <td>Expected: If existing tokens had the groupPrincipals field set to null to be updated / omitted on upgrade</td>
            <td>Actual: When upgrading with existing tokens that have the group principal field set to null they remain after upgrade<br><br>Pass/Fail: ❌<code>FAIL</code></td>
        </tr>
    </tbody>
</table>

Also noticed https://github.com/rancher/rancher/issues/39828 while validating this issue

**Comment by crobby** [comment]
For the reopened repro, was the token created while on 2.6.x?  If so, it would have been created with the nil groupPrincipals still, which would still present the same issue later on.  The fix for 2.7.0 was NOT backported, so old tokens still have the underlying issue.  This is expected.

**Comment by brudnak** [comment]
### ✅ All Checks Passing / Expected
Updated original comment above and closing issue
